### ISSUE TITLE
Exchange voices command is available with a single or list selection but does nothing

### ISSUE THREAD

**Issue by wizofaus** [musescore/MuseScore#23330]
### Issue type

UX/Interaction bug (incorrect behaviour)

### Bug description

If you select one or more notes in a list selection (using ctrl+click etc.), the various "exchange voices" commands from Tools|Voices are all enabled but do nothing.
Either the command should be disabled or they should work as expected (ideally the latter!).  

### Steps to reproduce

1. Create a score with a single note in voice 1 and a single note in voice 2
2. Select them both using ctrl+click (list selection, not range)
3. Choose "Tools|Voices|Exchange Voice 1&2"
4. Result: nothing happens
Expected: either 3. should not be possible because command is disabled, or better, it should work.

### MuseScore Version

4.3.1

### Regression

I don't know

### Operating system

Windows 11

**Comment by wizofaus** [comment]
Happy to fix either way, though the "actually work" option would take somewhat longer than the others. There's no message currently for "Please select a range" (or similar), so would require translation etc.

**Comment by irwir** [comment]
There should have been `Steps 5`:
...
5. Reading the handbook.

Apparently, the function name is dubious.
`Exchange voices` in MuseScore requires that one of the voices in the pair was empty.

The reasons for certain design decisions are unclear.
Preferable would be straitforward implementation as in `Swap voices` function of Sibelius.

**Comment by wizofaus** [comment]
Not sure what reading the handbook has to do with expected behaviour.  No command should ever be available and active but do nothing with no information about why nothing happened
And, no I use it all the time on passages with notes in both voices, it's not true that one must be empty at all.  If it says that in the Handbook then it's evidence that reading the handbook is more often than not a waste of time.  The source code is the only source of truth!

**Comment by irwir** [comment]
The handbook is still recommended; quoting:

> These commands work only on full measures (or multiple measures), not on individual notes.

So it was my misreading of your description (plus some false memories).

As for your suggestion, _Tools - Voices_ item could be disabled if selection is not full/multiple measures.

**Comment by wizofaus** [comment]
Then the handbook is still wrong because it works fine if you select a range that's not a full measure too (though I haven't done extensive testing to see if it handles all scenarios).

**Comment by MarcSabatella** [comment]
If a partial measure is selected, the command still operates on the full measure.  That’s how it handles all scenarios.  it’s easy with full measures, often next to impossible otherwise.

**Comment by wizofaus** [comment]
I'd say that's a separate bug then, though I'd fix at the same time

**Comment by MarcSabatella** [comment]
It’s not a bug, it’s by design as the consensus choice (as of some years ago) from among the several possible ways of handling partial measures.  Certainly, though, alternative designs for the handling of partial measures could potentially be proposed, discussed, agreed upon, implemented, and documented.

**Comment by wizofaus** [comment]
My proposal:
if you have a selection is that is NOT a range of 1 or more whole ranges then:
a) if doing exchange A-B, first take all the selected notes in voice A and do the equivalent of "Ctrl+Alt+n" to assign to voice 5 (normally not accessible, but useful as a temporary holding voice)
b) take all notes selected in voice B and do the equivalent of Ctrl+Alt+<A> to assign them to voice A
c) take all notes in voice 5 and do the equivalent of Ctrl+Alt+B to assign them to voice B

Otherwise, use existing algorithm.

**Comment by MarcSabatella** [comment]
As discussed in Discord, it's just not that simple.  If the voices don't start or end at the same time position, there is no guarantee that you'll be able to do the moves, and you'll end up either doing nothing for some of the notes, or stuck in places where some note has nowhere to go.

Here was one example I noted:

![image](https://github.com/musescore/MuseScore/assets/1799936/d1a5dad5-efcf-4444-a53c-46629899856c)

Just one of many many possible cases to consider.  Your algorithm above will leave the third selected "C" nowhere to go, except maybe back where it started (try it using voice 3 as the temp voice and see).  Or you can modify the rhythm to chop the start/end notes up and only move the portions they have in common, but there is no reason to suppose a user would actually expect that either.

I'm not saying it's hard to do *something*, but almost anything you choose is going to be surprising and almost impossible to document.  And not obviously better than the current solution.

So as I said, alternative designs for the handling of partial measures could potentially be proposed, discussed, agreed upon, implemented, and documented, but acknowledging the complexity of the problem I(from a user understanding perspective - not from a code perspective) is definitely part of that.

**Comment by wizofaus** [comment]
Ok so that brings up another piece of (to me) very unexpected behaviour, that probably deserves a separate bug report, but currently if you do select a note in, say, voice 2 and there's also a note at the same place already in voice 1, then using "Ctrl+Alt+1" or clicking on "voice 1" button in the toolbar does nothing at all with no explanation.  I'd prefer to fix *that* first (again, it's simple in principle - it should delete material to make room for the note, same as changing a note duration etc.).
But if there's some reason not to fix that, then sure, exchange voices should give a message if a) the selection is anything other than a range of 1 or more complete measures and b) it's impossible to assign some of the notes to another voice based on the current limitations of that command.
On the hand I don't have an issue with the fact that in many cases, if you have a selection that isn't entire measures, performing "exchange 1-2" twice may or may not return you to what it was before (because material has been deleted etc.).

**Comment by MarcSabatella** [comment]
This too is by design, the consensus choice as to the best way of handling such cases.  Again, it’s possible to propose, discuss, agree upon, implement, and document a new design.

**Comment by irwir** [comment]
> Your algorithm above will leave the third selected "C" nowhere to go, except maybe back where it started

As a fun fact, some OCR/PDF conversters actually managed to create voices longer than time signature defines.
MuseScore complained on every save.

**Comment by wizofaus** [comment]
Also, I took your example from above, changed the time signature to 3/8, used exchange voices on the 3 selected notes, then changed back to 4/4:

![image](https://github.com/musescore/MuseScore/assets/5364225/b14b02b8-7ea3-4cb7-b475-93b88b2a68a9)

Not that it proves anything, but does demonstrate that the current solution is not necessarily ideal.

**Comment by irwir** [comment]
> Also, I took your example from above, changed the time signature to 3/8, used exchange voices on the 3 selected notes, then changed back to 4/4:

IInitial selection with two quarters in lower voice you converted to quater and eighth.
This is bad cheating.

**Comment by wizofaus** [comment]
Exactly, you shouldn't have to resort to that to try to get a useful result. Not that I ever would try to use exchange voices on such a selection. But it's definitely a command that would be useful to use on selections that aren't entire measures. I'm happy to put up a PR so you can play around with it, will probably have to wait till tomorrow or the next day though.
